lib/elixir_sense/core/compiler.ex

```diff
@@ -0,0 +1,4696 @@
+        {name, _, atom} when is_atom(name) and is_atom(atom) ->
+          normalized = normalize_rescue(e_right, e)
+
+          match_context =
+            for exception <- normalized, reduce: nil do
+              nil -> {:struct, [], {:atom, exception}, nil}
+              other -> {:union, [other, {:struct, [], {:atom, exception}, nil}]}
+            end
+
+          match_context =
+            if match_context == nil do
+              {:struct, [], {:atom, Exception}, nil}
+            else
+              match_context
+            end
+
+          vars_with_inferred_types = TypeInference.find_typed_vars(e_left, match_context, :match)
+          sr = State.merge_inferred_types(sr, vars_with_inferred_types)
+
+          {{:in, meta, [e_left, normalized]}, sr, er}
+
+        _ ->
+          # elixir rejects this case, we normalize to underscore
+          {{:in, meta, [{:_, [], e.module}, normalize_rescue(e_right, e)]}, sr, er}
+      end
+    end
+
+    # rescue expr() => rescue expanded_expr()
+    defp expand_rescue({_, meta, _} = arg, s, e) do
+      # TODO how to check for cursor here?
+      case ElixirExpand.Macro.expand_once(arg, %{e | line: ElixirUtils.get_line(meta)}) do
+        ^arg ->
+          # elixir rejects this case
+          # try to recover from error by generating fake expression
+          expand_rescue({:in, meta, [arg, {:_, [], e.module}]}, s, e)
+
+        new_arg ->
+          expand_rescue(new_arg, s, e)
+      end
+    end
+
+    # rescue list() or atom() => _ in (list() or atom())
+    defp expand_rescue(arg, s, e) do
+      expand_rescue({:in, [], [{:_, [], e.module}, arg]}, s, e)
+    end
+
+    defp normalize_rescue(atom, _e) when is_atom(atom) do
+      [atom]
+    end
+
+    defp normalize_rescue(other, e) do
+      # elixir is strict here, we reject invalid nodes
+      res =
+        if is_list(other) do
+          Enum.filter(other, &is_atom/1)
+        else
+          []
+        end
+
+      if res == [] do
+        [{:_, [], e.module}]
+      else
+        res
+      end
+    end
+
+    defp expand_clauses(fun, {key, [_ | _] = clauses}, s, e) do
+      transformer = fn clause, sa ->
+        {e_clause, s_acc, _e_acc} =
+          clause(fun, clause, new_vars_scope(sa), e)
+
+        {e_clause, remove_vars_scope(s_acc, sa)}
+      end
+
+      {values, se} = Enum.map_reduce(clauses, s, transformer)
+      {{key, values}, se}
+    end
+
+    defp expand_clauses(fun, {key, expr}, s, e) do
+      # try to recover from error by wrapping the expression in a clauses list
+      # elixir raises here bad_or_missing_clauses
+      expand_clauses(fun, {key, [expr]}, s, e)
+    end
+
+    # helpers
+
+    defp sanitize_opt(opts, opt) do
+      case Keyword.fetch(opts, opt) do
+        :error -> []
+        {:ok, value} -> [{opt, value}]
+      end
+    end
+
+    defp sanitize_opts(opts, allowed) do
+      Enum.flat_map(allowed, fn opt -> sanitize_opt(opts, opt) end)
+    end
+  end
+
+  defmodule Bitstring do
```

Comment:
We can probably simplify the bitstring implementation. A lot of the work here is to normalize the bitstrings, which you don't care. You would probably be happy enough by simply traversing the left side of `::` (or the entry if there is no `::`).

Reply:
Sure. Here I care only for the parts of tree that recurse into expand or define variables

Reply:
Yeah, variables can only be defined on the left side, so you are good!